src/zarr/buffer.py

```diff
@@ -0,0 +1,157 @@
+            return Endian.big
+        else:
+            return Endian(sys.byteorder)
+
+    def reshape(self, newshape: Iterable[int]) -> Self:
+        return self.__class__(self.as_numpy_array().reshape(newshape))
+
+    def astype(self, dtype: np.DTypeLike, order: Literal["K", "A", "C", "F"] = "K") -> Self:
+        return self.__class__(self.as_numpy_array().astype(dtype=dtype, order=order))
+
+    def __getitem__(self, key: Any) -> Self:
+        # print("__getitem__: \n", np.asanyarray(self.as_numpy_array().__getitem__(key)))
+        return self.__class__(np.asanyarray(self.as_numpy_array().__getitem__(key)))
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        if isinstance(value, NDBuffer):
+            value = value.as_numpy_array()
+        self.as_numpy_array().__setitem__(key, value)
+
+    def __len__(self) -> int:
+        return self.as_numpy_array().__len__()
+
+    def fill(self, value: Any) -> None:
+        self.as_numpy_array().fill(value)
+
+    def copy(self) -> Self:
+        return self.__class__(self.as_numpy_array().copy())
+
+    def transpose(self, *axes: np.SupportsIndex) -> Self:
+        return self.__class__(self.as_numpy_array().transpose(*axes))
+
+
+class Buffer(NDBuffer):
+    """Contiguous memory block
+
+    We use `Buffer` throughout Zarr to represent a contiguous block of memory.
+    For now, we only support host memory but the plan is to support other types
+    of memory such as CUDA device memory.
+    """
+
+    @classmethod
+    def create_empty(
+        cls, *, shape: Iterable[int], dtype: np.DTypeLike = "b", order: Literal["C", "F"] = "C"
+    ) -> Self:
+        return cls(np.empty(shape=shape, dtype=dtype, order=order))
+
+    def memoryview(self) -> memoryview:
+        return memoryview(self._data.reshape(-1).view(dtype="b"))
+
+    def as_numpy_array(self, dtype: Optional[np.DTypeLike] = "b") -> np.ndarray:
+        return self._data.reshape(-1).view(dtype=dtype)
+
+    def to_bytes(self) -> bytes:
+        return bytes(self.memoryview())
+
+    def __getitem__(self, key: Any) -> Self:
+        return self.__class__(self.as_numpy_array().__getitem__(key))
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        self.as_numpy_array().__setitem__(key, value)
+
+    def __len__(self) -> int:
+        return self._data.nbytes
+
+    def __add__(self, other: Buffer) -> Self:
+        return self.__class__(np.frombuffer(self.to_bytes() + other.to_bytes(), dtype="b"))
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, (bytes, bytearray)):
+            return self.to_bytes() == other
+        raise ValueError(
+            f"equal operator not supported between {self.__class__} and {other.__class__}"
+        )
+
+
+def as_nd_buffer(data: Any) -> NDBuffer:
+    if isinstance(data, NDBuffer):
+        return data
+    return NDBuffer(np.asanyarray(data))
+
+
+def as_ndarray(data: Optional[NDBuffer]) -> Optional[np.ndarray]:
```

Comment:
I think it might be good to keep data conversion explicit in the zarr-python codebase. That is, I think all components (stores, codecs, etc.) should be forced to use a method that makes the data type explicit. This will make the code clearer by showing intent and avoid unnecessary implicit data conversion. E.g., if a component access `__cuda_array_interface__` to see if the buffer is on device, the buffer might have to move data from host to device in order to provide said `__cuda_array_interface__`. 

But I agree, the user facing API should support the different protocols implicitly but let’s address this in a follow-up PR?

Reply:
So, we would want to add a `.to(device)` method to the `NDBuffer` to do H2D and D2H transfers down the line and then how would this function change? Something like this?
```python
def as_ndarray(data: Optional[NDBuffer], device: Optional[Device]) -> Optional[Union[np.ndarray, cp.ndarray]]:
```

Reply:
Yes, something like `.to(device: str)` where `device` is either a predefined location like `host` or `cuda`, or some new user-defined location that each user-provided component in the pipeline understand. The idea is to make it as flexible as possible -- an user should be able to plug-in a whole custom made pipeline.

---

I was thinking more like:
```python
def as_ndarray(
    data: Optional[NDBuffer], 
    factory_function: Callable[[NDBuffer], ArrayLike] = np.asarray
) -> Optional[ArrayLike]:
```
where `factory_function` should return an array given a `NDBuffer`. 

Now, this is where it would be helpful if `NDBuffer` implements the different buffer protocols like DLPack, `__array_interface__`, `__cuda_array_interface__`, etc. This way, the factory function can just be `np.asarray`, `cupy.asarray`, or `torch.tensor` in most cases.